token_provider.go

```diff
@@ -25,13 +26,17 @@ type Token interface {
 type ManualRefreshTokenProvider struct {
 	tokenClient *TokenClient
 	token       Token
+	tokenMutex  sync.Mutex
```

Comment:
Use sync.RWMutex instead. Otherwise, all requests using the same client will need to acquire the lock before running, effectively creating a queue.